File: ios/Cove/CoinControlManager.swift
import os
import SwiftUI

extension WeakReconciler: CoinControlManagerReconciler where Reconciler == CoinControlManager {}

private enum CoinControlManagerError: LocalizedError {
    case closed

    var errorDescription: String? {
        "Coin control manager is closed"
    }
}

@Observable final class CoinControlManager: ReconcilingManager, CoinControlManagerReconciler {
    typealias Message = CoinControlManagerReconcileMessage
    typealias Action = CoinControlManagerAction

    private struct RustState {
        var rust: RustCoinControlManager?
        var isClosed = false
    }

    private let logger = Log(id: "CoinControlManager")
    @ObservationIgnored
    private let rustState = OSAllocatedUnfairLock(initialState: RustState())
    @ObservationIgnored
    private let rustBridge = DispatchQueue(label: "cove.CoinControlManager.rustbridge", qos: .userInitiated)

    let id: WalletId

    private(set) var sort: CoinControlListSort? = .some(.date(.descending))

    var search: String = ""
    var totalSelected = Amount.fromSat(sats: 0)
    var selected: Set<Utxo.ID> = []
    var utxos: [Utxo]
    var lockStateLoadFailed: Bool
    var unit: Unit = .sat

    private var updateSendFlowManagerTask: Task<Void, Never>? = nil

    private var rust: RustCoinControlManager? {
        rustState.withLock { $0.rust }
    }

    @ObservationIgnored
    var searchBinding: Binding<String> {
        Binding(
            get: { self.search },
            set: {
                if self.search != $0 { self.dispatch(.notifySearchChanged($0)) }
                self.search = $0
            }
        )
    }

    @ObservationIgnored
    var selectedBinding: Binding<Set<Utxo.ID>> {
        Binding(
            get: { self.selected },
            set: {
                let visibleIds = Set(self.utxos.map(\.id))
                let visibleSpendableIds = Set(self.utxos.filter(\.spendable).map(\.id))
                let selectedOutsideVisibleSearch = self.selected.subtracting(visibleIds)
                let selected = selectedOutsideVisibleSearch.union($0.intersection(visibleSpendableIds))

                self.selected = selected
                self.dispatch(.notifySelectedUtxosChanged(Array(selected)))
            }
        )
    }

    public init(_ rust: RustCoinControlManager) {
        self.id = rust.id()

        self.utxos = rust.utxos()
        self.lockStateLoadFailed = rust.lockStateLoadFailed()
        self.unit = rust.unit()

        rustState.withLock { $0.rust = rust }
        rust.listenForUpdates(reconciler: WeakReconciler(self))
    }

    deinit {
        close()
    }

    func close() {
        guard takeRustForClose() != nil else { return }

        logger.debug("Closing CoinControlManager")
        updateSendFlowManagerTask?.cancel()
        updateSendFlowManagerTask = nil
    }

    private func takeRustForClose() -> RustCoinControlManager? {
        rustState.withLock { state in
            guard !state.isClosed else { return nil }

            state.isClosed = true
            let rust = state.rust
            state.rust = nil
            return rust
        }
    }

    public func buttonArrow(_ key: CoinControlListSortKey) -> String? {
        _ = self.sort
        guard let rust else { return nil }

        return switch rust.buttonPresentation(button: key) {
        case .selected(.ascending):
            "arrow.up"
        case .selected(.descending):
            "arrow.down"
        case .notSelected:
            .none
        }
    }

    public func isSortSelected(_ key: CoinControlListSortKey) -> Bool {
        guard let rust else { return false }

        if case .selected = rust.buttonPresentation(button: key) { return true }
        return false
    }

    public func selectedUtxos() -> [Utxo] {
        rust?.selectedUtxos() ?? []
    }

    public var totalSelectedAmount: String {
        displayAmount(self.totalSelected)
    }

    public var totalSelectedSats: Int {
        Int(self.totalSelected.asSats())
    }

    public func continuePressed() {
        guard let sfm = AppManager.shared.sendFlowManager else { return }
        self.updateSendFlowManagerTask?.cancel()
        self.updateSendFlowManagerTask = nil

        sfm.dispatch(.setCoinControlMode(selectedUtxos()))
    }

    private func updateSendFlowManager() {
        guard let sfm = AppManager.shared.sendFlowManager else { return }
        self.updateSendFlowManagerTask?.cancel()
        self.updateSendFlowManagerTask = Task {
            try? await Task.sleep(for: .milliseconds(100))
            guard !Task.isCancelled else { return }
            sfm.dispatch(.setCoinControlMode(selectedUtxos()))
        }
    }

    var canApplyReconcileMessages: Bool {
        rust != nil
    }

    func apply(_ message: Message) {
        switch message {
        case let .updateSort(sort):
            withAnimation { self.sort = sort }
        case .clearSort:
            withAnimation { self.sort = .none }
        case let .updateUtxos(utxos):
            withAnimation { self.utxos = utxos }
        case let .updateSearch(search):
            withAnimation { self.search = search }
        case let .updateSelectedUtxos(utxos: selected, totalSelected):
            updateSendFlowManager()
            self.selected = Set(selected)
            withAnimation { self.totalSelected = totalSelected }
        case let .updateUnit(unit):
            withAnimation { self.unit = unit }
        case let .updateLockStateLoadFailed(failed):
            withAnimation { self.lockStateLoadFailed = failed }
        }
    }

    func displayAmount(_ amount: Amount, showUnit: Bool = true) -> String {
        switch (unit, showUnit) {
        case (.btc, true):
            amount.btcStringWithUnit()
        case (.btc, false):
            amount.btcString()
        case (.sat, true):
            amount.satsStringWithUnit()
        case (.sat, false):
            amount.satsString()
        }
    }

    func reloadLabels() async {
        guard let rust else { return }

        await rust.reloadLabels()
    }

    func setSpendability(_ spendable: Bool, for outpoint: OutPoint) async throws {
        guard let rust else { throw CoinControlManagerError.closed }

        try await rust.setUtxoSpendability(outpoint: outpoint, spendable: spendable)
    }

    func logReconcile(message: Message) {
        logger.debug("reconcile: \(message)")
    }

    func logReconcileMany(messages: [Message]) {
        logger.debug("reconcile_messages: \(messages)")
    }

    public func dispatch(action: Action) {
        dispatch(action)
    }

    public func dispatch(_ action: Action) {
        rustBridge.async { [weak self] in
            guard let self, let rust = self.rust else { return }

            self.logger.debug("dispatch: \(action)")
            rust.dispatch(action: action)
        }
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/CoinControlManager.kt
package org.bitcoinppl.cove

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.bitcoinppl.cove_core.*
import org.bitcoinppl.cove_core.types.*
import java.io.Closeable
import java.util.concurrent.atomic.AtomicBoolean

/**
 * coin control manager - manages UTXO selection for send flow
 * ported from iOS CoinControlManager.swift
 */
@Stable
class CoinControlManager internal constructor(
    private val rust: RustCoinControlManager,
) : CoinControlManagerReconciler,
    Closeable {
    private val tag = "CoinControlManager"

    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val isClosed = AtomicBoolean(false)
    private val rustGuard =
        RustHandleGuard(
            ownerName = "CoinControlManager",
            handleName = "RustCoinControlManager",
            isClosed = isClosed,
        ) {
            android.util.Log.w(tag, it)
        }

    val id: WalletId = rust.id()

    var sort by mutableStateOf<CoinControlListSort?>(
        CoinControlListSort.Date(ListSortDirection.DESCENDING),
    )
        private set

    var search by mutableStateOf("")

    var totalSelected by mutableStateOf(Amount.fromSat(0u))
        private set

    var selected by mutableStateOf<Set<ULong>>(emptySet())

    var utxos by mutableStateOf<List<Utxo>>(emptyList())
        private set

    var lockStateLoadFailed by mutableStateOf(false)
        private set

    var unit by mutableStateOf(BitcoinUnit.SAT)
        private set

    private var updateSendFlowManagerTask: Job? = null

    init {
        logDebug("Initializing CoinControlManager")
        utxos = rust.utxos()
        lockStateLoadFailed = rust.lockStateLoadFailed()
        unit = rust.unit()
        rust.listenForUpdates(this)
    }

    private fun logDebug(message: String) {
        android.util.Log.d(tag, message)
    }

    private fun <T> withRust(
        block: RustCoinControlManager.() -> T,
    ): T = rustGuard.withHandle(rust, block)

    private fun <T> withRustOr(
        defaultValue: T,
        block: RustCoinControlManager.() -> T,
    ): T = rustGuard.withHandleOr(rust, defaultValue, block)

    private suspend fun <T> withRustSuspend(
        block: suspend RustCoinControlManager.() -> T,
    ): T = rustGuard.withHandleSuspend(rust, block)

    private suspend fun <T> withRustOrSuspend(
        defaultValue: T,
        block: suspend RustCoinControlManager.() -> T,
    ): T = rustGuard.withHandleOrSuspend(rust, defaultValue, block)

    /**
     * update search and dispatch notification
     */
    fun updateSearch(value: String) {
        if (search != value) {
            dispatch(CoinControlManagerAction.NotifySearchChanged(value))
        }
        search = value
    }

    /**
     * update selected utxos and dispatch notification
     */
    fun updateSelected(value: Set<ULong>) {
        val visibleIds = utxos.map { it.outpoint.hashToUint() }.toSet()
        val visibleSpendableIds = utxos.filter { it.spendable }.map { it.outpoint.hashToUint() }.toSet()
        val selectedOutsideVisibleSearch = selected.subtract(visibleIds)
        val spendableSelection = selectedOutsideVisibleSearch.union(value.intersect(visibleSpendableIds))
        selected = spendableSelection

        val hiddenSelectedOutpoints =
            withRustOr(emptyList()) {
                selectedUtxos()
            }
                .filter { selectedOutsideVisibleSearch.contains(it.outpoint.hashToUint()) }
                .map { it.outpoint }
        val visibleSelectedOutpoints =
            utxos
                .filter { it.spendable && spendableSelection.contains(it.outpoint.hashToUint()) }
                .map { it.outpoint }
        val outpoints = hiddenSelectedOutpoints + visibleSelectedOutpoints
        dispatch(CoinControlManagerAction.NotifySelectedUtxosChanged(outpoints))
    }

    /**
     * get current button presentation based on sort state
     */
    fun buttonPresentation(key: CoinControlListSortKey): ButtonPresentation? =
        withRustOr(null) {
            buttonPresentation(key)
        }

    /**
     * get button arrow icon based on sort state
     */
    fun buttonArrow(key: CoinControlListSortKey): String? =
        when (val presentation = buttonPresentation(key)) {
            is ButtonPresentation.Selected -> {
                when (presentation.v1) {
                    ListSortDirection.ASCENDING -> "arrow_upward"
                    ListSortDirection.DESCENDING -> "arrow_downward"
                }
            }
            is ButtonPresentation.NotSelected -> null
            null -> null
        }

    val totalSelectedAmount: String
        get() = displayAmount(totalSelected)

    val totalSelectedSats: Long
        get() = totalSelected.asSats().toLong()

    /**
     * called when user presses continue button
     * navigates forward to CoinControlSetAmount screen with selected UTXOs
     */
    fun continuePressed(app: AppManager) {
        val walletId = id
        val selectedUtxos =
            withRustOr(emptyList()) {
                selectedUtxos()
            }

        // navigate forward to coin control set amount screen
        val sendRoute = SendRoute.CoinControlSetAmount(walletId, selectedUtxos)
        app.pushRoute(Route.Send(sendRoute))
    }

    private fun updateSendFlowManager() {
        val sfm = AppManager.getInstance().sendFlowManager ?: return
        updateSendFlowManagerTask?.cancel()
        updateSendFlowManagerTask =
            mainScope.launch {
                delay(SEND_FLOW_UPDATE_DELAY_MS)
                if (!isActive) return@launch

                val selectedUtxos =
                    withRustOr(emptyList()) {
                        selectedUtxos()
                    }
                sfm.dispatch(SendFlowManagerAction.SetCoinControlMode(selectedUtxos))
            }
    }

    private fun apply(message: CoinControlManagerReconcileMessage) {
        when (message) {
            is CoinControlManagerReconcileMessage.UpdateSort -> {
                sort = message.v1
            }

            is CoinControlManagerReconcileMessage.ClearSort -> {
                sort = null
            }

            is CoinControlManagerReconcileMessage.UpdateUtxos -> {
                utxos = message.v1
            }

            is CoinControlManagerReconcileMessage.UpdateSearch -> {
                search = message.v1
            }

            is CoinControlManagerReconcileMessage.UpdateSelectedUtxos -> {
                updateSendFlowManager()
                selected = message.utxos.map { it.hashToUint() }.toSet()
                totalSelected = message.totalValue
            }

            is CoinControlManagerReconcileMessage.UpdateUnit -> {
                unit = message.v1
            }

            is CoinControlManagerReconcileMessage.UpdateLockStateLoadFailed -> {
                lockStateLoadFailed = message.v1
            }
        }
    }

    fun displayAmount(amount: Amount, showUnit: Boolean = true): String =
        when (unit to showUnit) {
            BitcoinUnit.BTC to true -> amount.btcStringWithUnit()
            BitcoinUnit.BTC to false -> amount.btcString()
            BitcoinUnit.SAT to true -> amount.satsStringWithUnit()
            BitcoinUnit.SAT to false -> amount.satsString()
            else -> amount.satsStringWithUnit()
        }

    suspend fun reloadLabels() {
        withRustOrSuspend(Unit) {
            reloadLabels()
        }
    }

    suspend fun setSpendability(
        outpoint: OutPoint,
        spendable: Boolean,
    ) = withRustSuspend {
        setUtxoSpendability(outpoint, spendable)
    }

    override fun reconcile(message: CoinControlManagerReconcileMessage) {
        if (isClosed.get()) return

        logDebug("reconcile: $message")
        mainScope.launch { apply(message) }
    }

    override fun reconcileMany(messages: List<CoinControlManagerReconcileMessage>) {
        if (isClosed.get()) return

        logDebug("reconcile_messages: ${messages.size} messages")
        mainScope.launch { messages.forEach { apply(it) } }
    }

    fun dispatch(action: CoinControlManagerAction) {
        logDebug("dispatch: $action")
        mainScope.launch(Dispatchers.IO) {
            withRustOr(Unit) {
                dispatch(action)
            }
        }
    }

    override fun close() {
        rustGuard.closeOnce {
            logDebug("Closing CoinControlManager")
            updateSendFlowManagerTask?.cancel()
            mainScope.cancel()
            rust.close()
        }
    }

    companion object {

        /**
         * delay before propagating coin control selection to SendFlowManager
         *
         * batches rapid UTXO selection changes into a single update
         * prevents excessive dispatch and recomputation during multi-select
         */
        private const val SEND_FLOW_UPDATE_DELAY_MS = 100L
    }
}
